### PR TITLE
Add scope delimiter option

### PR DESCRIPTION
**Pull request by thosakwe**
Fixes the following issues:
* dart-lang/tools#308

### CODE REVIEW

File: lib/src/authorization_code_grant.dart [hunk 1]
```diff
@@ -99,10 +99,8 @@ class AuthorizationCodeGrant {
   /// [httpClient] is used for all HTTP requests made by this grant, as well as
   /// those of the [Client] is constructs.
   AuthorizationCodeGrant(
-          this.identifier,
-          this.authorizationEndpoint,
-          this.tokenEndpoint,
-          {this.secret, bool basicAuth: true, http.Client httpClient})
+      this.identifier, this.authorizationEndpoint, this.tokenEndpoint,
+      {this.secret, bool basicAuth: true, http.Client httpClient})
```
Comment (nex3) [inline lib/src/authorization_code_grant.dart]: Please don't include unrelated formatting changes.

File: lib/src/authorization_code_grant.dart [hunk 2]
```diff
@@ -125,8 +123,8 @@ class AuthorizationCodeGrant {
   /// query parameters provided to the redirect URL.
   ///
   /// It is a [StateError] to call this more than once.
-  Uri getAuthorizationUrl(Uri redirect, {Iterable<String> scopes,
-      String state}) {
+  Uri getAuthorizationUrl(Uri redirect,
+      {Iterable<String> scopes, String state, String delimiter: ' '}) {
```
Comment (nex3) [inline lib/src/authorization_code_grant.dart]: Document the delimiter parameter, and assign the default in the body rather than in the parameter list (putting it in the parameter list makes it impossible for users to express "use the default" without explicitly encoding that default).

**Review by nex3 — CHANGES_REQUESTED** [review]

File: lib/src/authorization_code_grant.dart (hunk 1 above)
Comment (thosakwe) [reply lib/src/authorization_code_grant.dart]: Ok, I won't. Sorry, didn't know.

File: lib/src/authorization_code_grant.dart (hunk 2 above)
Comment (thosakwe) [reply lib/src/authorization_code_grant.dart]: Will do.

**Comment by thosakwe** [comment]
@nex3 Sorry for not updating this in a long time, but I've now made the requested changes.

**Comment by nex3** [comment]
One more thing (sorry I didn't catch this in the first review!): I think for `AuthorizationCodeGrant` and `Credentials`, the delimiter should be set in the constructor rather than in the method. A given endpoint is always going to want the same delimiter, and for `Credentials` in particular you need the `Client` to be able to call `refresh()` without any extra knowledge of the quirks of the endpoint's API.

**Comment by thosakwe** [comment]
Okay, that works. I'll update it tomorrow, thanks!

**Comment by thosakwe** [comment]
Not really sure why the CLA thing went away, but I've signed it before.

**Comment by nex3** [comment]
It looks like GitHub thinks that https://github.com/dart-lang/oauth2/pull/17/commits/dd57b5916a4039207bbeac0dd30cb9f21dae4802 has an invalid author email. Did you commit with a different email than you normally use?

**Comment by thosakwe** [comment]
No, it's the same one; however, I recently bought a new computer, so maybe that has something to do with it.

It's times like these when I realize how little I actually know about Git, haha.

**Comment by nex3** [comment]
Looking at the commits locally, it looks like your most recent one has your email as the committer name, and `Dh6743tyre08` as your email address. [This doc](https://help.github.com/articles/setting-your-email-in-git/) should say how to fix that.

**Comment by thosakwe** [comment]
Okay, so I was able to change my credentials locally. Had to rewrite the commit history and change my e-mail as well. CLA approval is back, woohoo!

File: lib/src/authorization_code_grant.dart [hunk 3]
```diff
@@ -59,6 +59,11 @@ class AuthorizationCodeGrant {
   /// documentation.
   final Uri authorizationEndpoint;
 
+  /// The scope strings will be separated by the provided [delimiter]. This
+  /// defaults to `" "`, the OAuth2 standard, but some APIs (such as Facebook's)
+  /// use non-standard delimiters.
+  final String delimiter;
```
Comment (nex3) [inline lib/src/authorization_code_grant.dart]: This should be private, and the constructor parameter should be documented independently.

File: lib/src/authorization_code_grant.dart [hunk 4]
```diff
@@ -148,7 +153,7 @@ class AuthorizationCodeGrant {
     };
 
     if (state != null) parameters['state'] = state;
-    if (!scopes.isEmpty) parameters['scope'] = scopes.join(' ');
+    if (!scopes.isEmpty) parameters['scope'] = scopes.join(delimiter ?? ' ');
```
Comment (nex3) [inline lib/src/authorization_code_grant.dart]: It would simplify the data model to assign the field to `' '` if the parameter isn't passed, rather than checking on every use.

File: lib/src/handle_access_token_response.dart [hunk 5]
```diff
@@ -22,11 +22,14 @@ const _expirationGrace = const Duration(seconds: 10);
...
 /// OAuth2 flow.
+/// 
+/// The scope strings will be separated by the provided [delimiter] (default: `' '`).
 Credentials handleAccessTokenResponse(
     http.Response response,
     Uri tokenEndpoint,
     DateTime startTime,
-    List<String> scopes) {
+    List<String> scopes,
+    {String delimiter}) {
```
Comment (nex3) [inline lib/src/handle_access_token_response.dart]: It's not worth making this optional since it's only called internally.

File: lib/src/resource_owner_password_grant.dart [hunk 6]
```diff
@@ -56,13 +61,13 @@ Future<Client> resourceOwnerPasswordGrant(
     }
   }
 
-  if (scopes != null && !scopes.isEmpty) body['scope'] = scopes.join(' ');
+  if (scopes != null && !scopes.isEmpty) body['scope'] = scopes.join(delimiter ?? ' ');
```
Comment (nex3) [inline lib/src/resource_owner_password_grant.dart]: Similarly to the above, just do `delimiter ??= ' '` at the top of the function.

**Review by nex3 — CHANGES_REQUESTED** [review]

File: lib/src/resource_owner_password_grant.dart (hunk 6 above)
Comment (thosakwe) [reply lib/src/resource_owner_password_grant.dart]: Okay, just pushed commits to correct these. Let me know if they're up to your expectations!

File: lib/src/authorization_code_grant.dart [hunk 7]
```diff
@@ -148,7 +156,7 @@ class AuthorizationCodeGrant {
     };
 
     if (state != null) parameters['state'] = state;
-    if (!scopes.isEmpty) parameters['scope'] = scopes.join(' ');
+    if (!scopes.isEmpty) parameters['scope'] = scopes.join(_delimiter ??= ' ');
```
Comment (nex3) [inline lib/src/authorization_code_grant.dart]: The `??` shouldn't be necessary here, since `_delimiter` is always non-null now.

File: lib/src/handle_access_token_response.dart [hunk 8]
```diff
@@ -78,7 +81,7 @@ Credentials handleAccessTokenResponse(
   }
 
   var scope = parameters['scope'] as String;
-  if (scope != null) scopes = scope.split(" ");
+  if (scope != null) scopes = scope.split(delimiter ?? " ");
```
Comment (nex3) [inline lib/src/handle_access_token_response.dart]: You shouldn't need the `??` here, since all callers are passing in non-null delimiters

**Review by nex3 — CHANGES_REQUESTED** [review]

**Comment by thosakwe** [comment]
Made the changes, but I'm an idiot and keep running into the CLA thing. Give me 2 minutes to fix it. Hopefully I'll have it done by the time you see this.

**Comment by nex3** [comment]
The code looks good! Can you add tests?

**Comment by thosakwe** [comment]
Sure. Which tests?

**Comment by nex3** [comment]
We should test every place the scope parameter escapes the package.

**Comment by thosakwe** [comment]
All right, I'll add some, and then just let me know if I'm missing anything!

**Review by nex3 — APPROVED** [review]

**Comment by nex3** [comment]
Looks good, thanks for iterating on this!

**Comment by thosakwe** [comment]
Awesome, thanks for the merge! This should fix dart-lang/tools#308. Now I'll get started on dart-lang/tools#303.
